ocaml/xapi/vgpuops.ml

```diff
@@ -34,135 +35,130 @@ let vgpu_of_vgpu ~__context vm_r vgpu =
     devid = int_of_string vgpu_r.API.vGPU_device;
     other_config = vgpu_r.API.vGPU_other_config;
     type_ref = vgpu_r.API.vGPU_type;
+    requires_passthrough = Xapi_vgpu.requires_passthrough ~__context ~self:vgpu;
   }
 
 let vgpus_of_vm ~__context vm_r =
   List.map (vgpu_of_vgpu ~__context vm_r) vm_r.API.vM_VGPUs
 
-let create_passthrough_vgpu ~__context ~vm vgpu available_pgpus pcis =
-  debug "Creating passthrough VGPUs";
+let allocate_vgpu_to_gpu ~__context host vgpu =
+  let available_pgpus = Db.Host.get_PGPUs ~__context ~self:host in
   let compatible_pgpus = Db.GPU_group.get_PGPUs ~__context ~self:vgpu.gpu_group_ref in
   let pgpus = List.intersect compatible_pgpus available_pgpus in
+  (* Sort the pgpus in lists of equal optimality for vGPU placement based on
+   * the GPU groups allocation algorithm *)
+  let sort_desc =
+    match Db.GPU_group.get_allocation_algorithm ~__context ~self:vgpu.gpu_group_ref with
+    | `depth_first -> false
+    | `breadth_first -> true
+  in
+  let sorted_pgpus = Helpers.sort_by_schwarzian ~descending:sort_desc
+      (fun pgpu ->
+         Helpers.call_api_functions ~__context (fun rpc session_id ->
+             Client.Client.PGPU.get_remaining_capacity ~rpc ~session_id
+               ~self:pgpu ~vgpu_type:vgpu.type_ref))
+      pgpus
+  in
   let rec choose_pgpu = function
     | [] -> None
     | pgpu :: remaining ->
       try
         Xapi_pgpu_helpers.assert_capacity_exists_for_VGPU_type ~__context
           ~self:pgpu ~vgpu_type:vgpu.type_ref;
-        Some (pgpu, Db.PGPU.get_PCI ~__context ~self:pgpu)
+        Some pgpu
       with _ -> choose_pgpu remaining
   in
-  match choose_pgpu pgpus with
-  | None ->
+  choose_pgpu sorted_pgpus
+
+let fail_creation vm vgpu =
+  match vgpu.requires_passthrough with
+  | None | Some `VF ->
+    raise (Api_errors.Server_error (Api_errors.vm_requires_vgpu, [
+        Ref.string_of vm;
+        Ref.string_of vgpu.gpu_group_ref;
+        Ref.string_of vgpu.type_ref
+      ]))
+  | Some `PF ->
     raise (Api_errors.Server_error (Api_errors.vm_requires_gpu, [
         Ref.string_of vm;
         Ref.string_of vgpu.gpu_group_ref
       ]))
-  | Some (pgpu, pci) ->
+
+let create_vgpu ~__context ~vm ~host vgpu =
```

Comment:
As discussed, this is just a wrapper function now (and a misleadingly named one at that), and its functionality can be incorporated into `allocate_vgpu_to_gpu` which is not called from anywhere else. The `vm` is implicit in the `vgpu`.